Lib/modulefinder.py

```diff
@@ -43,6 +52,87 @@ def ReplacePackage(oldname, newname):
     replacePackageMap[oldname] = newname
 
 
+def _find_module(name, path=None):
+    """An importlib reimplementation of imp.find_module (for our purposes)."""
+
+    # To correctly find the module on the given path,  we need to
+    # temporarily clear the import state of the program and modify sys.path.
```

Comment:
While this is a true statement, it's also substantial scope creep for `modulefinder`, which has historically only worked for regular filesystem paths. So I think it would be good to break this aspect of the change up into two parts:

1. In this PR, use [importlib.machiner.PathFinder.find_spec](https://docs.python.org/3/library/importlib.html#importlib.machinery.PathFinder) *without* clearing any of the import system caches to emulate the `imp.find_module` behaviour

2. File a follow-up enhancement request to expand module finder to respect `sys.meta_path` hooks

Reply:
I realised I left out why the change concerns my and I'd like to see it reviewed separately: the actual `imp.find_module` implementation doesn't have any side effects, so calling it from multiple threads is safe. Rebinding `sys` module attributes on the other hand is decidedly *un*safe, so if that enhancement is going to happen, we'll need to find a different way of doing it that doesn't have global side effects :)

Reply:
Thanks for the feedback @ncoghlan. I've replaced the use of `importlib.util.find_spec` with `importlib.machinery.PathFinder.find_spec` for now.

On concurrency - I agree that it is a valid concern here. It would be trivial to include the proposed changes inside of a `threading.Lock` if we keep the `sys` modifications, but I can break that all out into a separate issue + PR if you feel it needs more discussion/review. I personally see `sys.meta_path`-respecting behavior as more of a "fix" than an "enhancement", but I'll defer to your judgement here.